### PR TITLE
Bif descriptor

### PR DESCRIPTION
**Pull request by AlexeySpizhevoy**
This PR adds an implementation of bio-inspired features (BIF) from the paper: Guo, Guodong, et al. "Human age estimation using bio-inspired features."  Computer Vision and Pattern Recognition, 2009. CVPR 2009.

It's added into module 'face'.

### CODE REVIEW

**Comment by AlexeySpizhevoy** [comment]
@vpisarev please take a look

File: modules/face/include/opencv2/face/bif.hpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+     *  @param image Input image (CV_32FC1).
+     *  @param features Feature vector (CV_32FC1).
+     */
+    CV_WRAP virtual void compute(InputArray image,
+                                 OutputArray features) const = 0;
+};
+
+/**
+ * @param num_bands The number of filter bands (<=12) used for computing BIF.
+ * @param num_rotations The number of image rotations for computing BIF.
```
Comment (berak) [inline modules/face/include/opencv2/face/bif.hpp]: number of filter bands (<=8)  , number of image rotations (<=12)  , right ?

**Comment by berak** [comment]
this looks nice !

can we maybe have a small usage example ? (does it need pre / post processing, what kind of classifier should be used)

(unfortunately, the paper seems to be behind a paywall)

File: modules/face/include/opencv2/face/bif.hpp (hunk 1 above)
Comment (AlexeySpizhevoy) [inline modules/face/include/opencv2/face/bif.hpp]: fixed

**Comment by AlexeySpizhevoy** [comment]
@berak thanks! It's just a descriptor, not a complete method. Images can be preprocessed if necessary, and any suitable ML method, in principle, can be used (SVM, decision forests, etc).

An example usage is described in the paper. Pdf can be found via scholar.google.com. Here's [one](http://vision.lbl.gov/Conferences/cvpr/Papers/data/papers/1186.pdf).

**Comment by berak** [comment]
thanks a lot for the link ! (i probably did not search hard enough)

**Comment by AlexeySpizhevoy** [comment]
Does anyone have any comments? Could this PR be merged?

**Comment by alalek** [comment]
@AlexeySpizhevoy Please squash all commits into one

**Comment by mshabunin** [comment]
- Aren't _bioinspired_ or _xfeatures2d_ modules more appropriate for this algorithm?
- Is it possible to inherit _BIF_ class from [cv::Feature2D](http://docs.opencv.org/master/d0/d13/classcv_1_1Feature2D.html#gsc.tab=0) class?

**Comment by vpisarev** [comment]
@AlexeySpizhevoy, thanks for the PR!
I think, as long as you squash all the commits into one, we can merge it! :+1:

@mshabunin, I believe, since it's for the face recognition, face is more approriate module; regarding `Feature2D` - its `compute()` method computes descriptors for the provided keypoints, whereas this algorithm computes descriptor for the specified image, i.e. the whole image acts as a "single keypoint". If we had something like "ImageDescriptor" or "GlobalFeature2D", we could derive from it, but for now the implemented solution is adequate

**Comment by AlexeySpizhevoy** [comment]
Squashed the commits.

**Pull request merged**